server/src/main/java/org/opensearch/search/aggregations/bucket/FastFilterRewriteHelper.java

```diff
@@ -354,86 +423,118 @@ public DateFieldMapper.DateFieldType getFieldType() {
 
-        if (ctx.reader().hasDeletions()) return false;
+        private final ValuesSourceConfig config;
+        private final Range[] ranges;
 
-        PointValues values = ctx.reader().getPointValues(fastFilterContext.fieldName);
-        if (values == null) return false;
-        // only proceed if every document corresponds to exactly one point
-        if (values.getDocCount() != values.size()) return false;
-
-        NumericDocValues docCountValues = DocValues.getNumeric(ctx.reader(), DocCountFieldMapper.NAME);
-        if (docCountValues.nextDoc() != NO_MORE_DOCS) {
-            logger.debug(
-                "Shard {} segment {} has at least one document with _doc_count field, skip fast filter optimization",
-                fastFilterContext.context.indexShard().shardId(),
-                ctx.ord
-            );
-            return false;
+        public RangeAggregationType(ValuesSourceConfig config, Range[] ranges) {
+            this.config = config;
+            this.ranges = ranges;
         }
 
-        // even if no ranges built at shard level, we can still perform the optimization
-        // when functionally match-all at segment level
-        if (!fastFilterContext.rangesBuiltAtShardLevel && !segmentMatchAll(fastFilterContext.context, ctx)) {
+        @Override
+        public boolean isRewriteable(Object parent, int subAggLength) {
+            if (config.fieldType() == null) return false;
+            MappedFieldType fieldType = config.fieldType();
+            if (fieldType.isSearchable() == false || !(fieldType instanceof PointFieldType)) return false;
+
+            if (parent == null && subAggLength == 0 && config.script() == null && config.missing() == null) {
+                if (config.getValuesSource() instanceof ValuesSource.Numeric.FieldData) {
+                    // ranges are already sorted by from and then to
+                    // we want ranges not overlapping with each other
+                    double prevTo = ranges[0].getTo();
+                    for (int i = 1; i < ranges.length; i++) {
+                        if (prevTo > ranges[i].getFrom()) {
+                            return false;
+                        }
+                        prevTo = ranges[i].getTo();
+                    }
+                    return true;
+                }
+            }
             return false;
         }
-        long[][] ranges = fastFilterContext.ranges;
-        if (ranges == null) {
-            logger.debug(
-                "Shard {} segment {} functionally match all documents. Build the fast filter",
-                fastFilterContext.context.indexShard().shardId(),
-                ctx.ord
-            );
-            ranges = fastFilterContext.buildRanges(ctx);
-            if (ranges == null) {
-                return false;
+
+        @Override
+        public Ranges buildRanges(SearchContext ctx, MappedFieldType fieldType) {
```

Comment:
We are always operating on PointFieldTypes correct? Can we pull this up into the interface?

Reply:
This buildRanges method is also used by DateHistogramAggregator, so being general now. Will try to see if better way in the follow up refactoring PR.

Reply:
I put down a point in the PR description about the follow up refactor. We can pick that up later.